### PR TITLE
Hardening/3109 remove payload word part 2

### PR DESCRIPTION
**Pull request by kzangeli**
* Removed an unnecessary lookup of rest service.
+ Split connectionTreat (MHD callback) in three functions

### CODE REVIEW

File: src/lib/common/string.cpp [hunk 1]
```diff
@@ -162,7 +162,7 @@ bool getIPv6Port(const std::string& in, std::string& outIp, std::string& outPort
 *
 * stringSplit - 
 */
-int stringSplit(const std::string& in, char delimiter, std::vector<std::string>& outV)
+int stringSplit(const std::string& in, char delimiter, std::vector<std::string>& outV, bool skipLastComponentIfEmpty)
```
Comment (fgalan) [inline src/lib/common/string.cpp]: New parameter (skipLastComponentIfEmpty) is a bit obscure... An explantation (or example) in the function preamble would be great.

File: src/lib/jsonParse/jsonRequest.cpp [hunk 2]
```diff
@@ -198,14 +198,13 @@ std::string jsonTreat
 
   if (reqP == NULL)
   {
-    std::string  details = std::string("Sorry, no request treating object found for RequestType /") + requestType(request) + "/";
     std::string  errorReply;
     char         reqTypeV[STRING_SIZE_FOR_INT];
 
-    restErrorReplyGet(ciP, SccBadRequest, details, &errorReply);
+    restErrorReplyGet(ciP, SccBadRequest, "service not found", &errorReply);
```
Comment (fgalan) [inline src/lib/jsonParse/jsonRequest.cpp]: Maybe "service not found" should be a #define? I remember we have a .h with the "magic strings" used in error messages.

**Comment by kzangeli** [comment]
this part was not even in the initial plan.
just something I "left" for later as it was far from trivial to implement all this, when I implemented the earlier search of the rest service.

About splitting the connectionTreat function in three separate functions ... that I believe might be an old issue

File: src/lib/rest/RestService.cpp [hunk 3]
```diff
@@ -69,7 +69,7 @@ RestService*                     restBadVerbV          = NULL;
 
 /* *****************************************************************************
 *
-* restServiceGet -
+* restServiceVectorGet -
```
Comment (fgalan) [inline src/lib/rest/RestService.cpp]: This refactor is hard to follow... could you summarize the changes a little bit, pls?

File: src/lib/common/string.cpp (hunk 1 above)
Comment (kzangeli) [reply src/lib/common/string.cpp]: I thought the name was descriptive enough ...
It's about not understanding /xxx/ as two components, but one (skip last component if it is empty).
but sure, it's easy enough to comment this in the function header

File: src/lib/rest/restServiceLookup.cpp [hunk 4]
```diff
@@ -97,5 +92,28 @@ RestService* restServiceLookup(ConnectionInfo* ciP, bool* badVerbP)
...
+  // But only one recursive call !
+  //
+
+  if (*badVerbP == false)
+  {
+    *badVerbP = true;
+    return restServiceLookup(ciP, badVerbP);
+  }
+
+  return &restBadVerbV[104];  // FIXME 02: 99.9% sure we never get here, but if ... Change the 104 service for a fixed one
```
Comment (fgalan) [inline src/lib/rest/restServiceLookup.cpp]: From where the magic number 104 did come?

File: src/lib/rest/RestService.cpp (hunk 3 above)
Comment (kzangeli) [reply src/lib/rest/RestService.cpp]: Well, this very function only changes name.
the problem comes later, in the following function.

what is changed is that i no longer lookup the rest service in the rest Service vector, as that is already done, and saved in ciP->restserviceP.

the diff is really messy ... :(

File: test/functionalTest/cases/0876_entity_dates/entity_dates_overridden_by_user_subs.test [hunk 5]
```diff
@@ -584,7 +584,7 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 =======================================
 POST http://localhost:REGEX(\d+)/notify
 Fiware-Servicepath: /
-Content-Length: 208
+Content-Length: REGEX((208|291))
```
Comment (fgalan) [inline test/functionalTest/cases/0876_entity_dates/entity_dates_overridden_by_user_subs.test]: Another weird change in Content-Length... This .test should be annnotated in the issue we created about it.

File: test/unittests/CMakeLists.txt [hunk 6]
```diff
@@ -203,6 +171,38 @@ SET (SOURCES
     rest/restReply_test.cpp
     rest/RestService_test.cpp
     rest/rest_test.cpp
+
+    # serviceRoutines/badVerbGetOnly_test.cpp
```
Comment (fgalan) [inline test/unittests/CMakeLists.txt]: Leftover?

Comment (fgalan) [reply test/unittests/CMakeLists.txt]: I undesrtand the test has been disabled maybe due to the effort it would take to adjust them. In the past you mentioned they should be removed... If so, let's do that in this PR, no need to wait any longer.

File: test/unittests/rest/RestService_test.cpp [hunk 7]
```diff
@@ -136,7 +136,7 @@ TEST(RestService, payloadParse)
 *
 * noSuchService -
 */
-TEST(RestService, noSuchServiceAndNotFound)
+TEST(RestService, DISABLED_noSuchServiceAndNotFound)
```
Comment (fgalan) [inline test/unittests/rest/RestService_test.cpp]: Maybe the test should be enterily removed? Related with https://github.com/apinf/fiware-orion/pull/35/review_comment/create#r208706721

File: src/lib/rest/restServiceLookup.cpp (hunk 4 above)
Comment (kzangeli) [reply src/lib/rest/restServiceLookup.cpp]: yeah, this is beyond ugly.
I am sure we never get there, as all possible errors are caught before.
That said, `&restBadVerbV[104]` is a pointer to the last restservice item in the badVerb vector, the one that takes care of any error, not yet found. 

As the fixme says, should be changed for a *fixed* struct for this case.

fixme in place, not really in a hurry to implement it

File: test/functionalTest/cases/0876_entity_dates/entity_dates_overridden_by_user_subs.test (hunk 5 above)
Comment (kzangeli) [reply test/functionalTest/cases/0876_entity_dates/entity_dates_overridden_by_user_subs.test]: yes, this is one more for that strange thing that is happening ...

File: test/unittests/CMakeLists.txt (hunk 6 above)
Comment (kzangeli) [reply test/unittests/CMakeLists.txt]: no, i decided to leave all those.
might do the effort to put all these unit tests back in, soon.
just not right now

File: test/unittests/rest/RestService_test.cpp (hunk 7 above)
Comment (kzangeli) [reply test/unittests/rest/RestService_test.cpp]: sure, we can remove all of them, not that i care very much.
however ... is it really better to remove then than to keep them to perhaps some day (soon) incorporate again?
i decided to go conservative ...

**Comment by kzangeli** [comment]
And yes, we're still in step one. Stage 4 seems OK

File: src/app/contextBroker/orionRestServices.cpp [hunk 8]
```diff
@@ -510,7 +510,7 @@ static RestService badVerbV[] =
   { InvalidRequest,                                2, { "ngsi9",   "*"                                                                 }, badNgsi9Request           },
   { InvalidRequest,                                2, { "ngsi10",  "*"                                                                 }, badNgsi10Request          },
   { InvalidRequest,                                0, { "*", "*", "*", "*", "*", "*"                                                   }, badRequest                },
-  { InvalidRequest,                                0, {                                                                                }, NULL                      },
```
Comment (fgalan) [inline src/app/contextBroker/orionRestServices.cpp]: Does the refactor and changes done in this PR has any impact in Developers Manual?

File: test/functionalTest/cases/0876_entity_dates/entity_dates_overridden_by_user_subs.test (hunk 5 above)
Comment (kzangeli) [reply test/functionalTest/cases/0876_entity_dates/entity_dates_overridden_by_user_subs.test]: Note added in issue https://github.com/telefonicaid/fiware-orion/issues/3258

File: test/unittests/CMakeLists.txt (hunk 6 above)
Comment (kzangeli) [reply test/unittests/CMakeLists.txt]: NTC

File: test/unittests/rest/RestService_test.cpp (hunk 7 above)
Comment (kzangeli) [reply test/unittests/rest/RestService_test.cpp]: Again, prefer to leave it.
some day we will reincorporate all these functests

So, NTC

File: src/lib/rest/restServiceLookup.cpp (hunk 4 above)
Comment (kzangeli) [reply src/lib/rest/restServiceLookup.cpp]: Also, once the new way of getting the info for "bad verb" (step 3 in issue #3109) this will go away

File: src/lib/common/string.cpp (hunk 1 above)
Comment (kzangeli) [reply src/lib/common/string.cpp]: Fixed in 4dfd0c058

File: src/lib/jsonParse/jsonRequest.cpp (hunk 2 above)
Comment (kzangeli) [reply src/lib/jsonParse/jsonRequest.cpp]: Fixed in 4dfd0c0

File: src/app/contextBroker/orionRestServices.cpp (hunk 8 above)
Comment (kzangeli) [reply src/app/contextBroker/orionRestServices.cpp]: No, we didn't get down to this level of details in the manual, fortunately ...
